File: sdk/python/src/securenotify/managers/apikey_manager.py
# SPDX-License-Identifier: Apache-2.0
# Copyright (c) 2026 KirkyX. All rights reserved.

from typing import Optional, List, Dict, Any
from datetime import datetime

from securenotify.types.api import (
    ApiKeyCreateRequest,
    ApiKeyCreateResponse,
    ApiKeyInfo,
)
from securenotify.utils.helpers import parse_datetime
from .base import BaseManager


class ApiKeyManager(BaseManager):
    async def create(
        self, name: str, permissions: List[str], expires_in: Optional[int] = None
    ) -> ApiKeyCreateResponse:
        """Create a new API key.

        Raises:
            ValueError: If name or permissions is empty.
            SecureNotifyApiError: On API error.
        """
        request = ApiKeyCreateRequest(
            name=name, permissions=permissions, expires_in=expires_in
        )
        return await self._execute("create_api_key", request)

    async def get(self, key_id: str) -> ApiKeyInfo:
        data = await self._execute("get_api_key", key_id)
        return ApiKeyInfo(
            id=data["id"],
            key_prefix=data["key_prefix"],
            name=data["name"],
            permissions=data["permissions"],
            is_active=data["is_active"],
            created_at=parse_datetime(data, "created_at"),
            last_used_at=parse_datetime(data, "last_used_at"),
            expires_at=parse_datetime(data, "expires_at"),
        )

    async def list(self) -> List[ApiKeyInfo]:
        data = await self._execute("list_api_keys")
        keys = []
        for item in data.get("keys", []):
            keys.append(
                ApiKeyInfo(
                    id=item["id"],
                    key_prefix=item["key_prefix"],
                    name=item["name"],
                    permissions=item["permissions"],
                    is_active=item["is_active"],
                    created_at=parse_datetime(item, "created_at"),
                    last_used_at=parse_datetime(item, "last_used_at"),
                    expires_at=parse_datetime(item, "expires_at"),
                )
            )
        return keys

    async def revoke(self, key_id: str) -> Dict[str, Any]:
        return await self._execute("revoke_api_key", key_id)

File: sdk/java/src/main/java/securenotify/managers/ApiKeyManager.java
// SPDX-License-Identifier: Apache-2.0
// Copyright (c) 2026 KirkyX. All rights reserved.

package securenotify.managers;

import securenotify.types.ApiKeyInfo;
import securenotify.types.ApiRequest;
import securenotify.types.ApiResponse;
import securenotify.types.PaginationResult;
import securenotify.utils.HttpClient;
import securenotify.utils.RetryHandler;

import java.util.Map;

/**
 * Manager for API key operations.
 */
public class ApiKeyManager {

    private final HttpClient httpClient;
    private final RetryHandler retryHandler;

    public ApiKeyManager(HttpClient httpClient) {
        this(httpClient, RetryHandler.DEFAULT);
    }

    public ApiKeyManager(HttpClient httpClient, RetryHandler retryHandler) {
        this.httpClient = httpClient;
        this.retryHandler = retryHandler;
    }

    /**
     * @param expiresIn Optional expiration time in seconds
     * @return The created API key (including the full key)
     */
    public ApiKeyInfo.ApiKeyCreateResponse create(String name, String userId, String[] permissions, Integer expiresIn) throws Exception {
        ApiRequest<?> request = ApiRequest.builder()
                .name(name)
                .userId(userId)
                .permissions(permissions)
                .expiresIn(expiresIn)
                .build();

        return retryHandler.execute(() -> {
            try {
                return httpClient.post("api/keys", request, ApiKeyInfo.ApiKeyCreateResponse.class);
            } catch (Exception e) {
                throw new RuntimeException(e);
            }
        }).getData();
    }

    public ApiKeyInfo.ApiKeyCreateResponse create(String name, String userId, String[] permissions) throws Exception {
        return create(name, userId, permissions, 604800); // 7 days default
    }

    /**
     * @return The API key info (without the full key)
     */
    public ApiKeyInfo get(String keyId) throws Exception {
        return retryHandler.execute(() -> {
            try {
                return httpClient.get("api/keys/" + keyId, null, ApiKeyInfo.class);
            } catch (Exception e) {
                throw new RuntimeException(e);
            }
        }).getData();
    }

    public ApiKeyListResponse list(Integer limit, Integer offset) throws Exception {
        ApiRequest<?> request = ApiRequest.builder()
                .limit(limit)
                .offset(offset)
                .build();

        return retryHandler.execute(() -> {
            try {
                return httpClient.post("api/keys", request, ApiKeyListResponse.class);
            } catch (Exception e) {
                throw new RuntimeException(e);
            }
        }).getData();
    }

    public ApiKeyListResponse list() throws Exception {
        return list(50, 0);
    }

    public ApiResponse<?> revoke(String keyId, String reason) throws Exception {
        ApiRequest<?> request = ApiRequest.builder()
                .reason(reason)
                .build();

        return retryHandler.execute(() -> {
            try {
                return httpClient.delete("api/keys/" + keyId, request, Map.class);
            } catch (Exception e) {
                throw new RuntimeException(e);
            }
        });
    }

    public static class ApiKeyListResponse {
        private ApiKeyInfo[] keys;
        private PaginationResult pagination;

        public ApiKeyInfo[] getKeys() {
            return keys;
        }

        public void setKeys(ApiKeyInfo[] keys) {
            this.keys = keys;
        }

        public PaginationResult getPagination() {
            return pagination;
        }

        public void setPagination(PaginationResult pagination) {
            this.pagination = pagination;
        }
    }
}
